### ISSUE TITLE
cf restart causes downtime unexpectedly

### ISSUE THREAD

**Issue by bleach** [cloudfoundry/cli#922]
### Command

cf restart APPNAME

### CLI Version

cf version 6.21.0+dff2cf8-2016-07-27

### Error

HTTP Error from a request to an app URL during the time taken for the app instances to restart

```
404 Not Found: Requested route ('APPNAME.domainl') does not exist.
```

### Details

One of our users reports that the behaviour of `cf restart` caused downtime to them unexpectedly. Their app takes a minute or so to start, so there's about a minute when requests to their app get a 404 from the router. They were expecting CF to provide a rolling restart.

The help text doesn't mention that this will cause an outage and the user is not warned when running the command. In comparison `cf scale` does warn the user when changing per-instance limits, but doesn't explicitly say that there will be downtime.

## Possible solutions

- Some help text could be added to explain that `cf restart` stops all instances of the app then starts all instances of the app - it's not a rolling restart.
- Help text could explicitly state that application downtime will occur while the instances restart
- A warning could be added that prompts the user to acknowledge the downtime risk. (This has the downside of potentially breaking existing scripts)

**Comment by bleach** [comment]
I should have made it clear that I'm interested in resolving this problem myself, so please do state if you have a preference for which solution is implemented, even if you have no time to do so yourself :)

**Comment by dkoper** [comment]
Hi @bleach 

I believe the V3 API may allow an improvement in downtime. Until then, I favor the first option.
A warning in help like in `cf auth` and `cf login`.
We can't introduce a prompt that breaks scripts until we release a 7.0, which I assume your users don't want to wait for.
I suggest you let us implement it - lot's of refactoring going on at the moment so your PR will likely need regular rebasing. We can work on the wording together. We're currently working on improving help UX, so I'd like to implement it as part of this epic.
How about:
```
NAME:
   restart - Restart an app

USAGE:
   cf restart APP_NAME

WARNING:
   This operation stops all instances of the app and then starts them again, so there is a period of downtime.

ALIAS:
   rs
```

Regards,
Dies Koper
CF CLI PM

**Comment by bleach** [comment]
Hi @dkoper 

I agree with that and the wording seems fine to me.

Is there a place to file changes that break compatibility in readiness for a future major release?

**Comment by dcarley** [comment]
`cf restage` also has the same behaviour. Should it get the same treatment as `cf restart`?

**Comment by bleach** [comment]
Now that you've pointed that out, @dcarley, I realise `cf push` doesn't warn that it will cause downtime to an existing app either.

**Comment by dkoper** [comment]
@bleach We have an area in the Pivotal tracker where I place stories that break compatibility, for consideration for a future major release. You can submit issues in the tracker here: although I would like to close them here as a "won't fix" as to not to clutter the issue list, I'll keep the linked story open and place it in the "V7" area.

Restart is a convenience operation of combining "stop" and "start". You can't **re**start something without stopping it.
I'm open to improving the help output of commands (the current command descriptions of `restart` and `restage` in particular don't add any value), but feeling reluctant to add warnings everywhere for things that have been obvious to most people.
I'd like to see how the V3 API functionality changes/adds to the CLI workflows, especially if we are going to be able to natively support zero-downtime workflows, and whether they'd replace the current workflows or be in addition.

**Comment by dkoper** [comment]
We have improved the `restart` help text in the just published 6.22.2 release:
https://github.com/cloudfoundry/cli#downloads

Regards,
Dies Koper
CF CLI PM

**Comment by bleach** [comment]
Thanks Dies!
